File: day4/McpStreamableAuth/src/lib/auth-middleware.ts
import type { NextFunction, Request, Response } from "express";
import {
	SCALEKIT_CONFIG,
	scalekit,
	WWW_AUTHENTICATE_HEADER,
} from "./scalekit-config.js";

// Extend Express Request to include our custom properties
declare global {
	namespace Express {
		interface Request {
			token?: string;
			// biome-ignore lint/suspicious/noExplicitAny: JWT claims are provider-specific
			tokenClaims?: any;
			isAuthenticated?: boolean;
		}
	}
}

/**
 * Mandatory authentication middleware that requires valid Bearer tokens for all requests
 * Requests without valid tokens will be rejected with 401 Unauthorized
 */
export async function requiredAuthMiddleware(
	req: Request,
	res: Response,
	next: NextFunction,
): Promise<void> {
	try {
		// Extract Bearer token from Authorization header (scheme is case-insensitive per RFC 7235)
		const authHeader = req.headers.authorization;
		const token = authHeader?.toLowerCase().startsWith("bearer ")
			? authHeader.slice("bearer ".length).trim()
			: null;

		if (!token) {
			throw new Error("Bearer token required");
		}

		// Validate token signature, expiry, issuer, and audience.
		//
		// We intentionally do NOT pass `requiredScopes` here because different
		// tools need different scopes — gating everything at the middleware
		// would either lock out tools that don't need a scope or force us to
		// grant the union of all scopes to every caller.
		//
		// Instead we enforce scopes inside each tool handler via
		// `requireScopes(...)` from `auth-context.ts`. See
		// `pony_password_advanced` for a working example. To gate a scope
		// globally (e.g. "any caller must have `mcp:access`"), add
		// `requiredScopes: ["mcp:access"]` here — Scalekit will then 401 any
		// token missing those scopes before it ever reaches a tool handler.
		const claims = await scalekit.validateToken(token, {
			audience: [SCALEKIT_CONFIG.resourceId],
		});

		// Attach token and claims to request for use by tools
		req.token = token;
		req.tokenClaims = claims;
		req.isAuthenticated = true;
		console.log("✓ Authenticated request with token");

		next();
	} catch (err) {
		// Invalid or missing token - return 401 with WWW-Authenticate header
		console.warn("⚠️ Authentication failed:", err);
		res
			.status(401)
			.header(WWW_AUTHENTICATE_HEADER.key, WWW_AUTHENTICATE_HEADER.value)
			.end();
	}
}

File: day4-dotnet/McpStreamableAuth/AuthMiddleware.cs
using Microsoft.AspNetCore.Http;
using System.Net;
using System.IdentityModel.Tokens.Jwt;
using Microsoft.IdentityModel.Tokens;
using System.Security.Claims;

namespace McpStreamableAuth;

/// <summary>
/// Mandatory authentication middleware that requires valid Bearer tokens for all requests.
/// Requests without valid tokens will be rejected with 401 Unauthorized.
/// </summary>
public class RequiredAuthMiddleware
{
    private readonly RequestDelegate _next;
    private readonly ScalekitConfig _config;
    private readonly JwtSecurityTokenHandler _tokenHandler;

    public RequiredAuthMiddleware(RequestDelegate next, ScalekitConfig config)
    {
        _next = next;
        _config = config;
        _tokenHandler = new JwtSecurityTokenHandler();
    }

    public async Task InvokeAsync(HttpContext context)
    {
        try
        {
            // Extract Bearer token from Authorization header
            var authHeader = context.Request.Headers.Authorization.ToString();
            string? token = null;

            if (authHeader.StartsWith("Bearer ", StringComparison.OrdinalIgnoreCase))
            {
                token = authHeader.Substring("Bearer ".Length).Trim();
            }

            if (string.IsNullOrWhiteSpace(token))
            {
                throw new UnauthorizedAccessException("Bearer token required");
            }

            // Validate token using JWT validation
            // Note: This is a simplified validation. For production, you should:
            // 1. Fetch JWKS (JSON Web Key Set) from Scalekit's authorization server
            // 2. Validate the token signature using the public keys from JWKS
            // 3. Validate issuer, audience, expiration, etc.
            //
            // The TypeScript version uses scalekit.validateToken() which handles all of this.
            // For now, we do basic validation. In production, implement proper JWKS validation.
            
            var jwtToken = _tokenHandler.ReadJwtToken(token);
            
            // Validate audience (must match resource ID)
            var audience = jwtToken.Claims.FirstOrDefault(c => c.Type == "aud" || c.Type == ClaimTypes.NameIdentifier)?.Value;
            if (string.IsNullOrEmpty(audience) || 
                (audience != _config.ResourceId && audience != _config.Audience))
            {
                throw new SecurityTokenValidationException($"Invalid audience. Expected: {_config.ResourceId} or {_config.Audience}, Got: {audience}");
            }

            // Validate expiration
            if (jwtToken.ValidTo < DateTime.UtcNow)
            {
                throw new SecurityTokenExpiredException("Token has expired");
            }

            // Extract claims
            var claims = jwtToken.Claims.ToDictionary(c => c.Type, c => (object)c.Value);
            
            // Note: Signature validation is NOT performed here.
            // The Scalekit SDK's validateToken() in TypeScript validates the signature using JWKS.
            // For production, you should implement JWKS validation here.

            // Attach token and claims to HttpContext for use by tools
            context.Items["Token"] = token;
            context.Items["TokenClaims"] = claims;
            context.Items["IsAuthenticated"] = true;

            // Set auth context for the current async operation chain
            var authContext = new AuthContext
            {
                Token = token,
                TokenClaims = claims,
                IsAuthenticated = true
            };
            AuthContextAccessor.SetContext(authContext);

            Console.WriteLine("✓ Authenticated request with token");

            // Continue to next middleware
            await _next(context);
        }
        catch (Exception ex)
        {
            // Invalid or missing token - return 401 with WWW-Authenticate header
            Console.WriteLine($"⚠️ Authentication failed: {ex.Message}");
            context.Response.StatusCode = (int)HttpStatusCode.Unauthorized;
            context.Response.Headers.Append(WwwAuthenticateHeader.Key, 
                WwwAuthenticateHeader.GetValue(_config.ResourceId));
            await context.Response.WriteAsync("Unauthorized");
        }
    }
}

/// <summary>
/// Extension method to register the authentication middleware.
/// </summary>
public static class RequiredAuthMiddlewareExtensions
{
    public static IApplicationBuilder UseRequiredAuth(this IApplicationBuilder builder)
    {
        return builder.UseMiddleware<RequiredAuthMiddleware>();
    }
}

